File: showcase/agents/copilot-pr-assistant-hosted/src/agent.py
"""CopilotSDKAgent — run the GitHub Copilot SDK as a Foundry HOSTED agent.

Foundry's hosting adapter ``ResponsesHostServer`` accepts any object that
satisfies the **structural** ``SupportsAgentRun`` protocol (``id``/``name``/
``description`` + ``run()``/``create_session()``/``get_session()``) — it does NOT
require an Agent Framework ``Agent`` subclass. So this thin adapter wraps a
``copilot.CopilotClient`` session and exposes it as a hostable agent.

Mapping:
  • Foundry passes the conversation history in ``messages``; the Copilot session
    keeps its OWN history, so we forward only the latest user turn and rely on the
    per-session Copilot session for continuity.
  • ``run(stream=True)`` yields ``AgentResponseUpdate(contents=[Content.from_text])``
    bridged from Copilot ``assistant.message_delta`` events via an asyncio queue.
  • ``run(stream=False)`` returns ``AgentResponse(messages=[assistant Message])``.

HITL note: a Foundry hosted agent speaks the **Responses** protocol, not AG-UI, so
this hosted variant uses ``approve_all`` for tools. The human-in-the-loop
(``confirm_changes`` approval) lives in the self-hosted **gateway** variant where
AG-UI drives the approval card. See README.
"""
from __future__ import annotations

import asyncio
import logging
import os
from collections.abc import Awaitable, Mapping
from typing import Any
from uuid import uuid4

from agent_framework import AgentResponse, AgentResponseUpdate, AgentSession, Content, Message
from copilot import CopilotClient, ToolSet, define_tool
from copilot.session import PermissionHandler
from copilot.generated.session_events import (
    AssistantMessageData,
    AssistantMessageDeltaData,
    SessionIdleData,
)
from pydantic import BaseModel, Field

from provider import build_provider, MODEL_NAME

logger = logging.getLogger("copilot_hosted")

# ── Tools (mirror the gateway PR agent's read tools) ──────────────────────────
_CHANGED_FILES = [
    {"path": "src/auth/session.ts", "summary": "Add Secure + SameSite=Strict to the session cookie",
     "diff": "- res.cookie('sid', token, { httpOnly: true });\n+ res.cookie('sid', token, { httpOnly: true, secure: true, sameSite: 'strict' });"},
    {"path": "src/auth/session.test.ts", "summary": "Test the cookie security flags",
     "diff": "+ it('sets Secure + SameSite=Strict', () => { expect(setCookie).toMatch(/Secure/); });"},
]


class _DiffParams(BaseModel):
    path: str = Field(description="File path from list_changed_files")


def _build_tools() -> list[Any]:
    async def _list(_args: Any, _inv: Any) -> dict:
        return {"files": [{"path": f["path"], "summary": f["summary"]} for f in _CHANGED_FILES],
                "count": len(_CHANGED_FILES)}

    async def _diff(args: _DiffParams, _inv: Any) -> dict:
        f = next((x for x in _CHANGED_FILES if x["path"] == args.path), None)
        return {"path": f["path"], "diff": f["diff"]} if f else {"error": "no such file"}

    return [
        define_tool(name="list_changed_files",
                    description="List files changed on the working branch.",
                    handler=_list, skip_permission=True),
        define_tool(name="get_file_diff",
                    description="Get the unified diff for one changed file.",
                    handler=_diff, params_type=_DiffParams, skip_permission=True),
    ]


_SYSTEM = (
    "You are a senior engineer that reviews code changes and drafts pull requests. "
    "Call list_changed_files, then get_file_diff for each file, then summarize a clear "
    "PR title and description (what changed and why; mention security impact). Be concise."
)


class CopilotSDKAgent:
    """Adapter exposing a Copilot SDK session as a Foundry-hostable SupportsAgentRun."""

    def __init__(self) -> None:
        self.id: str = "copilot-pr-assistant-hosted"
        self.name: str | None = "Copilot PR Assistant (hosted)"
        self.description: str | None = "GitHub Copilot SDK agent, hosted on Foundry (Responses protocol)."
        self._client: CopilotClient | None = None
        self._sessions: dict[str, Any] = {}
        self._lock = asyncio.Lock()

    # ── SupportsAgentRun: sessions ────────────────────────────────────────────
    def create_session(self, *, session_id: str | None = None) -> AgentSession:
        return AgentSession(session_id=session_id)

    def get_session(self, service_session_id: str, *, session_id: str | None = None) -> AgentSession:
        return AgentSession(session_id=session_id, service_session_id=service_session_id)

    # ── internals ─────────────────────────────────────────────────────────────
    async def _ensure_client(self) -> CopilotClient:
        if self._client is None:
            self._client = CopilotClient(
                mode="empty",
                base_directory=os.environ.get("COPILOT_HOME", "/tmp/.copilot-home"),
                log_level="error",
            )
            await self._client.start()
        return self._client

    async def _copilot_session(self, key: str):
        async with self._lock:
            if key in self._sessions:
                return self._sessions[key]
            client = await self._ensure_client()
            sess = await client.create_session(
                session_id=f"foundry-{key}",
                model=MODEL_NAME,
                provider=await build_provider(),
                tools=_build_tools(),
                available_tools=ToolSet().add_custom("*"),
                system_message={"mode": "replace", "content": _SYSTEM},
                on_permission_request=PermissionHandler.approve_all,
                streaming=True,
            )
            self._sessions[key] = sess
            return sess

    @staticmethod
    def _latest_user_text(messages: Any) -> str:
        if isinstance(messages, str):
            return messages
        if not messages:
            return ""
        seq = messages if isinstance(messages, list) else [messages]
        for m in reversed(seq):
            role = getattr(m, "role", None)
            if str(role) == "user" or role == "user":
                for c in getattr(m, "contents", []) or []:
                    if getattr(c, "type", None) == "text" and getattr(c, "text", None):
                        return c.text
        # fall back to the last message's text
        last = seq[-1]
        for c in getattr(last, "contents", []) or []:
            if getattr(c, "text", None):
                return c.text
        return ""

    # ── SupportsAgentRun: run ─────────────────────────────────────────────────
    def run(
        self,
        messages: Any = None,
        *,
        stream: bool = False,
        session: AgentSession | None = None,
        function_invocation_kwargs: Mapping[str, Any] | None = None,
        client_kwargs: Mapping[str, Any] | None = None,
        **_: Any,
    ) -> Awaitable[AgentResponse] | Any:
        key = (session.service_session_id or session.session_id) if session else "default"
        key = key or "default"
        prompt = self._latest_user_text(messages)
        if stream:
            return self._run_stream(key, prompt)
        return self._run_once(key, prompt)

    async def _run_once(self, key: str, prompt: str) -> AgentResponse:
        sess = await self._copilot_session(key)
        chunks: list[str] = []

        def handler(event: Any) -> None:
            data = getattr(event, "data", None)
            if isinstance(data, AssistantMessageData) and data.content:
                chunks.append(data.content)

        unsub = sess.on(handler)
        try:
            await sess.send_and_wait(prompt, timeout=120.0)
        finally:
            unsub()
        text = chunks[-1] if chunks else ""
        return AgentResponse(messages=[Message(role="assistant", contents=[Content.from_text(text)])])

    async def _run_stream(self, key: str, prompt: str):
        sess = await self._copilot_session(key)
        queue: asyncio.Queue = asyncio.Queue()
        mid = uuid4().hex

        def handler(event: Any) -> None:
            data = getattr(event, "data", None)
            if isinstance(data, AssistantMessageDeltaData) and data.delta_content:
                queue.put_nowait(("delta", data.delta_content))
            elif isinstance(data, SessionIdleData):
                queue.put_nowait(("done", None))

        unsub = sess.on(handler)
        try:
            await sess.send(prompt)
            while True:
                kind, payload = await queue.get()
                if kind == "done":
                    break
                yield AgentResponseUpdate(
                    contents=[Content.from_text(payload)], role="assistant", message_id=mid,
                )
        finally:
            try:
                unsub()
            except Exception:  # noqa: BLE001
                pass

File: showcase/agents/copilot-pr-assistant/src/agent.ts
// The Copilot-SDK agent: a "Copilot PR Assistant".
//
// Copilot reads a (seeded) set of changed files, drafts a pull request, and calls
// `propose_pull_request`. That tool is NON-consequential here — it only records a
// DRAFT and returns immediately. The AG-UI bridge (bridge.ts) turns that draft
// into a human-in-the-loop `confirm_changes` gate; the PR is only actually opened
// after the user approves. This mirrors the forgewright HITL contract, but the
// engine is the GitHub Copilot SDK instead of Microsoft Agent Framework.
//
// mode:"empty" gives a bare runtime with ONLY our tools (no built-in file/shell
// tools), so the agent is safe to host and deterministic for a demo.

import { CopilotClient, defineTool, ToolSet, type Tool } from "@github/copilot-sdk";
import { DefaultAzureCredential, ManagedIdentityCredential } from "@azure/identity";
import { startApimAdapter } from "./apim-adapter.js";

export const AGENT_NAME = "copilot_pr_assistant";
export const PROPOSE_TOOL = "propose_pull_request";

// ── Seeded "repository" so Copilot has something concrete to summarize ─────────
const CHANGED_FILES = [
  {
    path: "src/auth/session.ts",
    summary: "Fix: session cookie missing `Secure` + `SameSite=Strict` flags",
    diff:
      "- res.cookie('sid', token, { httpOnly: true });\n" +
      "+ res.cookie('sid', token, { httpOnly: true, secure: true, sameSite: 'strict' });",
  },
  {
    path: "src/auth/session.test.ts",
    summary: "Add a test asserting the cookie security flags",
    diff:
      "+ it('sets Secure + SameSite=Strict on the session cookie', () => {\n" +
      "+   expect(setCookie).toMatch(/Secure/);\n" +
      "+   expect(setCookie).toMatch(/SameSite=Strict/);\n" +
      "+ });",
  },
];

// ── In-memory "GitHub" so the demo runs with no real side effects ──────────────
let _prCounter = 0;
export interface OpenedPr {
  number: number;
  url: string;
  title: string;
  base: string;
  head: string;
}
export function openPullRequest(p: { title: string; body: string; base: string; head: string }): OpenedPr {
  _prCounter += 1;
  return {
    number: _prCounter,
    url: `https://github.com/acme/widgets/pull/${_prCounter}`,
    title: p.title,
    base: p.base,
    head: p.head,
  };
}

// ── Tools (only these exist; mode:"empty") ─────────────────────────────────────
// Raw JSON-Schema params (not Zod) to avoid a bundled-zod type skew with the SDK.
export function buildTools(): Tool<any>[] {
  return [
    defineTool("list_changed_files", {
      description: "List the files changed on the working branch, with a one-line summary of each change.",
      parameters: { type: "object", properties: {}, required: [] },
      skipPermission: true,
      handler: async () => ({
        files: CHANGED_FILES.map((f) => ({ path: f.path, summary: f.summary })),
        count: CHANGED_FILES.length,
      }),
    }),
    defineTool("get_file_diff", {
      description: "Get the unified diff for a single changed file.",
      parameters: {
        type: "object",
        properties: { path: { type: "string", description: "File path from list_changed_files" } },
        required: ["path"],
      },
      skipPermission: true,
      handler: async ({ path }: { path: string }) => {
        const f = CHANGED_FILES.find((x) => x.path === path);
        return f ? { path: f.path, diff: f.diff } : { error: `No changed file at ${path}` };
      },
    }),
    // The "consequential" intent. It only DRAFTS — the bridge gates the real open.
    defineTool(PROPOSE_TOOL, {
      description:
        "Propose a pull request for the changes. This DRAFTS the PR and pauses for human approval; " +
        "it does NOT open the PR. Call this once you have a clear title/body. After calling it, STOP " +
        "and wait — do not call other tools or take further action.",
      parameters: {
        type: "object",
        properties: {
          title: { type: "string", description: "Concise PR title" },
          body: { type: "string", description: "PR description: what changed and why" },
          base: { type: "string", description: "Base branch", default: "main" },
          head: { type: "string", description: "Head branch", default: "fix/session-cookie-flags" },
        },
        required: ["title", "body"],
      },
      skipPermission: true,
      handler: async () =>
        "Draft recorded. Awaiting human approval before opening. Do not call any more tools; wait for the decision.",
    }),
  ];
}

// In mode:"empty" every session must opt into its tools explicitly. We expose
// ONLY our custom tools (no built-in file/shell tools) — safe + deterministic.
export function availableTools(): ToolSet {
  return new ToolSet().addCustom("*");
}

export const SYSTEM_MESSAGE =
  "You are a senior engineer that opens pull requests. Workflow:\n" +
  "1. Call list_changed_files, then get_file_diff for each file, to understand the change.\n" +
  "2. Draft a clear PR title and body (what changed and why; mention security impact).\n" +
  "3. Call propose_pull_request ONCE with the title/body. It pauses for HUMAN APPROVAL — " +
  "the PR is NOT opened until the human approves.\n" +
  "4. After calling propose_pull_request, STOP. Tell the user the PR is ready for their approval. " +
  "Do not invent a PR number or URL; the system assigns it after approval.";

// ── Model provider ─────────────────────────────────────────────────────────────
// Two key-management paths, in priority order:
//   1. APIM key  — set LLM_API_KEY + AZURE_OPENAI_ENDPOINT to an APIM gateway path.
//      The gateway fronts the model; the SDK sends the key as the Azure `api-key`
//      header. Simplest for shared demos behind an AI gateway.
//   2. Azure BYOM (keyless) — DefaultAzureCredential / ManagedIdentityCredential
//      bearer token (refreshed per session). Tokens expire ~1h.
// Either way: wireApi:"completions" (NOT "responses"), and a gpt-5/o-series model
// (the Copilot SDK encrypts prompts; gpt-4.x is unsupported).
function credential() {
  if (process.env.NODE_ENV === "development") return new DefaultAzureCredential();
  return process.env.AZURE_CLIENT_ID
    ? new ManagedIdentityCredential(process.env.AZURE_CLIENT_ID)
    : new ManagedIdentityCredential();
}

let _apimBaseUrl: string | null = null;

export async function buildProvider() {
  const apiVersion = process.env.AZURE_OPENAI_API_VERSION || "2025-04-01-preview";
  const wireApi = (process.env.WIRE_API as "completions" | "responses") || "completions";

  // Path A — APIM AI gateway (path-prefixed, subscription-key auth). The SDK can't
  // speak APIM's scheme directly, so we front it with an in-process auth adapter
  // and give the SDK a host-only `type:"azure"` endpoint.
  if (process.env.APIM_GATEWAY && process.env.APIM_KEY) {
    if (!_apimBaseUrl) {
      const adapter = await startApimAdapter({
        gateway: process.env.APIM_GATEWAY,
        pathPrefix: process.env.APIM_PATH_PREFIX || "",
        apiKey: process.env.APIM_KEY,
      });
      _apimBaseUrl = adapter.baseUrl;
    }
    return {
      type: "azure" as const,
      baseUrl: _apimBaseUrl,
      apiKey: "apim", // ignored by the adapter; it injects the real key
      wireApi,
      azure: { apiVersion },
    };
  }

  // Path B — a direct endpoint (Azure OpenAI host or OpenAI-compatible).
  const baseUrl = required("AZURE_OPENAI_ENDPOINT");
  const type = (process.env.PROVIDER_TYPE as "azure" | "openai") || "azure";
  const apiKey = process.env.LLM_API_KEY;
  if (apiKey) {
    return { type, baseUrl, apiKey, wireApi, azure: { apiVersion } };
  }
  const { token } = await credential().getToken("https://cognitiveservices.azure.com/.default");
  if (!token) throw new Error("Failed to acquire an Azure AD token for the model provider.");
  return { type, baseUrl, bearerToken: token, wireApi, azure: { apiVersion } };
}

export const MODEL_NAME = process.env.MODEL_NAME || "gpt-5.4-mini";

let _client: CopilotClient | null = null;
export async function getClient(): Promise<CopilotClient> {
  if (_client) return _client;
  // mode:"empty" → multi-user server mode with no built-in tools/instructions.
  // It requires an explicit per-session persistence location; baseDirectory sets
  // COPILOT_HOME, giving each sessionId its own state dir (also powers resume).
  const baseDirectory = process.env.COPILOT_HOME || `${process.cwd()}/.copilot-home`;
  _client = new CopilotClient({ mode: "empty", baseDirectory, logLevel: process.env.SDK_LOG_LEVEL as any || "error" });
  await _client.start();
  return _client;
}

function required(name: string): string {
  const v = process.env[name];
  if (!v) throw new Error(`${name} is required (Azure BYOM). See .env.example.`);
  return v.replace(/\/+$/, "");
}
